php_phongo.c

```diff
@@ -1382,11 +1382,21 @@ static const char* php_phongo_bson_type_to_string(bson_type_t type) /* {{{ */
 
 static bool php_phongo_uri_finalize_auth(mongoc_uri_t* uri) /* {{{ */
 {
+	const bson_t* credentials = mongoc_uri_get_credentials(uri);
+	bson_iter_t   iter;
+	const char*   source       = NULL;
+	const char*   username     = mongoc_uri_get_username(uri);
+	bool          require_auth = username != NULL;
+
+	if (bson_iter_init_find_case(&iter, credentials, MONGOC_URI_AUTHSOURCE)) {
+		source       = bson_iter_utf8(&iter, NULL);
+		require_auth = true;
```

Comment:
[DRIVERS-949](https://jira.mongodb.org/browse/DRIVERS-949) includes a test case where a username is _not_ required if `authSource` is `$external`. I don't think this accounts for that, so I'd suggest manually adding that test (since we can't use the auth spec tests directly) and revising this to relax validation.

While the username is optional for both MONGODB-X509 and MONGODB-IAM mechanisms, it looks like the test only cares about `authSource` since the mechanism is never specified.